### PR TITLE
`Development`: Minimize data transfer when processing a build result

### PR DESCRIPTION
**Pull request by krusche**
### Summary

Processing one build result issued 23 database statements and moved about 140 KB, almost all of it the exercise's problem statement and the course's code of conduct, which nothing on that path reads. It is now 12 statements and about 35 KB. Getting there meant replacing the ordered list of results, whose position carried the correction round, with an explicit column on the result, because that ordering was what forced every write in this area to load and re-save the whole submission.

### Motivation and Context

Follow-up to #13561. Measured on a build result with a 20 KB problem statement, a 4 KB grading instruction text, a 2 KB course description and an 8 KB code of conduct, four statements carried that free text and one row was 401 columns wide:

| # | Call site | Cols | Real bytes |
| --- | --- | ---: | ---: |
| 1 | the opening participation load | 83 | 24 728 |
| 7 | eager `participation` of the matched submission | **401** | 35 254 |
| 10 | `findByIdWithSubmissionPolicyElseThrow` | 48 | 24 307 |
| 15 | `save(submission)` merge-select | 220 | 25 078 |

Plus three separate `course` selects at roughly 10 KB each. At 2000 exam students each finishing one build that is around 280 MB per wave, and students push repeatedly.

The root cause of the widest rows is that `Submission.participation`, `Participation.exercise` and `Exercise.course` are all eager, so loading any submission entity loads all of it. The reason a submission entity had to be saved at all was the `@OrderColumn` on its results.

### Description

**Replacing the result ordering.** A submission's results were an ordered list whose position was the correction round, maintained by an `@OrderColumn`. Nothing read `results_order` other than the mapping itself, so the correction round was the only information the ordering carried. It becomes `result.correction_round`: 0 for the first correction, 1 for the second, null for automatic and Athena results, which are not correction rounds. The column is backfilled from the ordering it replaces, and `Submission.results` becomes an unordered `Set`. The newest result was already the one with the highest id, so nothing needed the position.

`SubmissionService.lockSubmission` sets the round from the round the tutor asked for, and `Submission.addResult` assigns the next one when a caller does not, which is the same moment the list position used to be decided.

**The API no longer emits null placeholders.** A tutor who had assessed only the second round produced a results array with a `null` at index 0, and the client indexed that array by correction round. `Result` now carries `correctionRound` and `getSubmissionResultByCorrectionRound` matches on it. Three components derived the round from `findIndex` over the results array and now read it off the result. `TextAssessmentService` no longer asserts with `!` that a result exists for the requested round; a round nobody has assessed simply has none.

**Reading only what is needed.** The submission a build result belongs to, and the submission policy, are read through projections rather than as entities.

**Writing only what changed.**
- The result owns the foreign key to its submission, so setting the submission before the insert writes it. This used to insert the result without its submission and then save the submission so its cascade filled the column in.
- Whether the build failed is one boolean on an existing row, written with a modifying query and only when it actually changed.
- Build log entries own their foreign key too, so they are saved with it, and a previous build's logs are deleted explicitly rather than through orphan removal on a collection. Their order comes from the timestamp they already carry instead of a position column.

**A footgun removed with it.** A test had to save the submission alongside a result because a result saved through its own repository kept `results_order` at its default of 0, collided with an existing result, and made the test fail non-deterministically. That cannot happen any more.

### Measurements

| Processing one build result | before | after |
| --- | ---: | ---: |
| Statements | 23 | **12** |
| Bytes moved | ~140 KB | **~35 KB** |
| Widest row | 401 columns | 83 |
| Statements carrying the free text | 4 | **1** |

### Steps for Testing

Prerequisites:
- 1 Instructor, 1 Student
- 1 Programming Exercise on a server with the integrated lifecycle setup (LocalVC and LocalCI)

1. As the student, push a passing commit. A result appears with the expected score.
2. Push a commit that does not compile. The submission is marked as build failed, the result scores 0, and the build logs are readable and in chronological order.
3. As the instructor, trigger all builds. A new result appears for the latest submission.
4. Push to the tests repository. The solution and template participations both get a result, and the exercise's test cases stay active.
5. As a tutor, open the exercise for assessment. The result created for the lock belongs to correction round 0, and the results array contains no null entries.
6. On an exam exercise with two correction rounds, assess the first and second rounds and check that each round shows its own result and that the dashboard's "continue assessment" link appears for the round it belongs to.
7. Configure an active lock repository policy with a limit of 1, exceed it, and trigger a build. The new result is unrated.

### Server Tests

1707 server tests were run locally across the assessment, submission, submission-policy, localci, architecture and integration suites, plus 265 client specs.

The migration's backfill has a PostgreSQL branch and a MySQL branch. CI only runs PostgreSQL, so the MySQL statement was run against a MySQL container with data covering automatic, Athena, manual, semi-automatic, null-assessment-type and orphan rows. Both branches produce identical output, matching the previous positional semantics, and re-running updates no rows.

#### Performance Review
- [x] I did not find any performance issues.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2

#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.tum.de/developer/guidelines/language).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.tum.de/developer/development-process#pr-naming-conventions).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.tum.de/developer/guidelines/performance) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.tum.de/developer/guidelines/server-development) and the [REST API guidelines](https://docs.artemis.tum.de/developer/guidelines/rest-api).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding guidelines](https://docs.artemis.tum.de/developer/guidelines/client-development).
- [x] I added multiple integration tests (Vitest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.tum.de/developer/guidelines/client-tests).
- [x] I documented the TypeScript code using JSDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **LocalVC** and **Jenkins**.

## Summary by CodeRabbit

* **New Features**
  * Added explicit correction-round tracking for manually assessed results.
  * Assessment views now select results by correction round rather than collection position.
  * Build logs are displayed in timestamp order with stable ordering.

* **Bug Fixes**
  * Improved handling of missing results during assessment workflows.
  * Corrected latest-result selection across multiple correction rounds.
  * Deleted results no longer contribute to course scores.
  * Preserved result ordering while preventing duplicate entries.

* **Refactor**
  * Submission results and programming build logs now use set-based collections.

### CODE REVIEW

File: src/main/java/de/tum/cit/aet/artemis/exercise/domain/Submission.java [hunk 1]
```diff
@@ -146,7 +148,7 @@ public Result getLatestResult() {
     @Nullable
     @JsonIgnore
     public Result getLatestCompletedResult() {
-        Result latestResult = Optional.ofNullable(results).orElse(List.of()).stream().filter(result -> result != null && result.getCompletionDate() != null)
+        Result latestResult = Optional.ofNullable(results).orElse(Set.of()).stream().filter(result -> result != null && result.getCompletionDate() != null)
                 .max(Comparator.comparing(Result::getCompletionDate)).orElse(null);
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/cit/aet/artemis/exercise/domain/Submission.java]: Fixed in 357d550115. `getLatestCompletedResult` now breaks the tie with the id comparator the other lookups on this class already use, so two results that share a completion date resolve to the same one on every request.

File: src/main/java/de/tum/cit/aet/artemis/exercise/service/SubmissionService.java [hunk 2]
```diff
@@ -183,8 +184,7 @@ public <T extends Submission> List<T> getAllSubmissionsAssessedByTutorForCorrect
         if (examMode) {
             var participations = this.studentParticipationRepository.findAllByParticipationExerciseIdAndResultAssessorAndCorrectionRoundIgnoreTestRuns(exerciseId, tutor);
             submissions = participations.stream().map(StudentParticipation::findLatestSubmission).filter(Optional::isPresent).map(Optional::get).map(submission -> (T) submission)
-                    .filter(submission -> submission.getResults().size() - 1 >= correctionRound && submission.getResults().get(correctionRound) != null)
-                    .collect(Collectors.toCollection(ArrayList::new));
+                    .filter(submission -> submission.hasResultForCorrectionRound(correctionRound)).collect(Collectors.toCollection(ArrayList::new));
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/cit/aet/artemis/exercise/service/SubmissionService.java:187]: Declining as out of scope. The gap is real but predates this PR: `findAllByParticipationExerciseIdAndResultAssessorAndCorrectionRoundIgnoreTestRuns` never filtered by correction round despite its name, and the code this PR replaced (`getResults().get(correctionRound) != null`) did not check the assessor of that round either. Tightening the assessor match changes exam correction-round semantics and needs its own PR with exam test coverage.

File: src/main/java/de/tum/cit/aet/artemis/programming/domain/ProgrammingSubmission.java [hunk 3]
```diff
@@ -41,9 +41,16 @@ public String getSubmissionExerciseType() {
...
+     * Only present if buildFailed == true.
+     * <p>
+     * Ordered by the time the build produced them rather than by a position column. A position column has to be
+     * maintained by Hibernate, which means every write has to go through this collection and therefore through a save of
+     * the whole submission, and saving a submission selects it together with its participation, exercise and course
+     * because those are eager. The entries carry their own timestamp, so the order does not need to be stored.
+     */
     @OneToMany(mappedBy = "programmingSubmission", cascade = CascadeType.ALL, orphanRemoval = true)
-    @OrderColumn(name = "build_log_entries_order")
+    @OrderBy("time")
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/cit/aet/artemis/programming/domain/ProgrammingSubmission.java]: Fixed in 357d550115: `@OrderBy("time, id")`. Two log lines written in the same millisecond are common, so the tie-breaker matters here. The client also sorts the entries by time now, so the displayed order does not rest on the mapping alone.

File: src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingSubmissionRepository.java [hunk 4]
```diff
@@ -76,6 +79,50 @@ default ProgrammingSubmission findFirstByParticipationIdAndCommitHashOrderByIdDe
...
+    @Query("""
+            SELECT new de.tum.cit.aet.artemis.programming.dto.BuildResultSubmissionDTO(
+                submission.id, submission.buildFailed, submission.commitHash, submission.type, submission.submissionDate,
+                submission.submitted, submission.exampleSubmission,
+                latestResult.id, latestResult.assessmentType, latestResult.completionDate, latestResult.correctionRound)
+            FROM ProgrammingSubmission submission
+                LEFT JOIN submission.results latestResult
+                    ON latestResult.id = (SELECT MAX(otherResult.id)
+                                          FROM submission.results otherResult)
+            WHERE submission.id = :submissionId
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingSubmissionRepository.java:101]: Declining. The query resolves a single submission by primary key, so the correlated subquery runs over that one submission's handful of results; an anti-join would not measurably change that and would rewrite a query that is covered by tests. Happy to revisit if a profile shows it mattering.

File: src/main/webapp/app/exercise/shared/entities/submission/submission.model.ts [hunk 5]
```diff
@@ -112,9 +112,16 @@ export function setSubmissionResultByCorrectionRound(submission: Submission, res
...
+    const existingIndex = submission.results.findIndex((existing) => existing?.correctionRound === correctionRound);
+    if (existingIndex >= 0) {
+        submission.results[existingIndex] = result;
+    } else {
+        submission.results = [...submission.results, result];
+    }
 
-    if (submission.results.length === correctionRound + 1) {
+    if (correctionRound === Math.max(...submission.results.map((existing) => existing?.correctionRound ?? -1))) {
         submission.latestResult = result;
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/webapp/app/exercise/shared/entities/submission/submission.model.ts:147]: Good catch, and the most important one on this PR. Fixed in 357d550115 on both sides. The server now orders the collection by id, so the serialized order is stable, and the client no longer reads a position: `getLatestSubmissionResult` delegates to a new `getNewestResult`, which picks the highest id and treats a result without an id as the newest. The two other places that took `results.last()` (`programming-exercise.service.ts` and `text-editor.component.ts`) go through the same helper. Covered by `submission.model.spec.ts`, which asserts the same data in both array orders resolves to the same result.

File: src/main/webapp/app/text/manage/assess/service/text-assessment.service.ts [hunk 6]
```diff
@@ -261,9 +261,13 @@ export class TextAssessmentService {
...
      *  @param result
      */
-    private static reconnectResultsParticipation(participation: Participation, submission: Submission, result: Result) {
+    private static reconnectResultsParticipation(participation: Participation, submission: Submission, result: Result | undefined) {
         setLatestSubmissionResult(submission, getLatestSubmissionResult(submission));
         submission.participation = participation;
+        // A correction round the tutor has not assessed yet simply has no result, so there is nothing to reconnect.
+        if (!result) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/webapp/app/text/manage/assess/service/text-assessment.service.ts:270]: Fixed in 357d550115: `loadFeedbackSuggestions` returns early when there is no result, since there is nothing to attach a suggestion to. Worth noting this one is pre-existing rather than introduced here, because the previous positional lookup also yielded a falsy result for an unassessed round.

File: src/main/webapp/app/localci/shared/entities/build-log.model.ts [hunk 7]
```diff
@@ -18,6 +18,19 @@ export type BuildLogEntry = {
...
+ * Compares two build log entries by their timestamp. Entries whose timestamp cannot be parsed keep their relative
+ * order, because sorting is stable.
+ */
+function compareByTime(first: BuildLogEntry, second: BuildLogEntry): number {
+    const firstTime = Date.parse(first.time);
+    const secondTime = Date.parse(second.time);
+    if (Number.isNaN(firstTime) || Number.isNaN(secondTime)) {
+        return 0;
+    }
+    return firstTime - secondTime;
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/webapp/app/localci/shared/entities/build-log.model.ts]: Correct, and worse than unstable: returning 0 for pairs involving an unparseable timestamp made the comparison intransitive, which leaves the result of a sort undefined. Fixed in 357d550115 by sorting only the entries with a usable timestamp among themselves and writing them back into the slots they occupied, so an unparseable entry keeps its position and never separates two entries that do compare. The interleaved case is now in `build-log.model.spec.ts`.

File: src/test/java/de/tum/cit/aet/artemis/exercise/participation/util/ParticipationUtilService.java [hunk 8]
```diff
@@ -1093,4 +1093,48 @@ private void ensureLocalVcRepositoryExists(URI repositoryUri) {
...
+        if (submission.getId() == null) {
+            // A submission that was never saved cannot have persisted results, and its collection is a plain one, so reading it is safe.
+            existing = submission.getResults().stream();
+        }
+        else {
+            existing = submissionRepository.findWithEagerResultsAndAssessorById(submission.getId()).map(Submission::getResults).orElse(Set.of()).stream();
+        }
+        // Compared by reference on purpose: the result is not saved yet, so its id is still null and equals would never match.
+        long manualResults = existing.filter(other -> other != null && other != result && other.getAssessmentType() != null && !other.isAutomatic() && !other.isAthenaBased())
+                .count();
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/test/java/de/tum/cit/aet/artemis/exercise/participation/util/ParticipationUtilService.java:1142]: Fixed in 357d550115: the filter now excludes the current result by id as well as by reference, so a caller passing a persisted result whose row the query returns as a different object no longer inflates the round.

File: src/main/java/de/tum/cit/aet/artemis/exercise/domain/Submission.java [hunk 9]
```diff
@@ -299,29 +302,49 @@ public Result getFirstManualResult() {
...
+     * and this is the same moment, so the behaviour is unchanged for every caller that does not set it itself.
+     * {@code SubmissionService.lockSubmission} does set it, from the round the tutor asked for, and that takes
+     * precedence. Automatic and Athena results are not correction rounds and keep no round.
      *
-     * @param result the {@link Result} which should be added
+     * @param result the result to add
      */
     public void addResult(Result result) {
+        if (result != null && result.getCorrectionRound() == null && !result.isAutomatic() && !result.isAthenaBased()) {
+            result.setCorrectionRound(countCorrectionRoundResults(result));
```
Comment (WoH) [inline src/main/java/de/tum/cit/aet/artemis/exercise/domain/Submission.java]: `addResult` must allocate `max(existing correctionRound) + 1`, not the count. after deleting round 0 from a submission that still has round 1, the next complaint result also gets round 1. `getResultForCorrectionRound(1)` then picks the older id, and round 2 copies stale feedback. add a fresh-session delete-gap test and a uniqueness constraint for non-null rounds.

File: src/main/webapp/app/exercise/shared/entities/submission/submission.model.ts [hunk 10]
```diff
@@ -58,7 +58,29 @@ export abstract class Submission implements BaseEntity {
  * @param submission
  */
 export function getLatestSubmissionResult(submission: Submission | undefined): Result | undefined {
-    return submission?.results?.last();
+    return getNewestResult(submission?.results);
```
Comment (WoH) [inline src/main/webapp/app/exercise/shared/entities/submission/submission.model.ts]: `getNewestResult` changes the converter input, but `setLatestSubmissionResult` still overwrites the final array slot. modeling history returns results by completion date descending, so `[200, 100]` becomes `[200, 200]` and result 100 disappears from feedback navigation. set only `latestResult` during response conversion, or replace an entry by id.

File: src/main/java/de/tum/cit/aet/artemis/exercise/service/SubmissionService.java [hunk 11]
```diff
@@ -629,6 +629,10 @@ else if (result == null) {
             result.setAssessor(userRepository.getUser());
         }
 
+        // The round this result belongs to is stored on the result itself. This is the one place where a manual result
+        // for a correction round is created or claimed, so it is also where a result that predates the column gets its
+        // round the first time a tutor opens it.
+        result.setCorrectionRound(correctionRound);
```
Comment (WoH) [inline src/main/java/de/tum/cit/aet/artemis/exercise/service/SubmissionService.java]: reject negative rounds before this write. `getResultForCorrectionRound` returns null for negative values, but this line persists the unchecked value. repeated `correction-round=-1` requests create hidden manual results and shift later round allocation. add a service-boundary check and a database check constraint.

File: src/main/resources/config/liquibase/changelog/20260825165553_changelog.xml [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        that area to load and re-save the whole submission so that Hibernate could maintain the column. On the path that
+        processes a build result that meant loading the submission together with its participation, exercise and course,
+        because those associations are eager, so the exercise's problem statement and the course's code of conduct
+        travelled over the wire for every single build.
+
+        Nothing reads results_order other than the mapping itself, so the correction round is the only piece of
+        information the ordering actually carried. It becomes a column on result: 0 for the first correction, 1 for the
+        second, null for automatic and Athena results, which are not correction rounds.
+
+        The column is left in place on purpose. Dropping it in the same release would make a rollback of the application
```
Comment (WoH) [inline src/main/resources/config/liquibase/changelog/20260825165553_changelog.xml]: retaining the column is not enough for binary rollback. the new set mapping stops writing `results_order`, so all post-upgrade results keep legacy default 0; the base list mapping then collapses duplicate slot 0 rows. direct build-log inserts likewise leave `build_log_entries_order` null. either dual-write both legacy columns for the rollback window or declare this migration forward-only and test upgrade, write, then downgrade.
